### PR TITLE
[Perf] Optimize hidden state extraction logic

### PR DESCRIPTION
**Pull request by benchislett**
## Purpose

Eliminate unnecessary synchronization and improve overlap compatibility of hidden state extraction. There is still some inefficiency due to the lack of full overlap, which should be further improved by https://github.com/vllm-project/vllm/pull/32951

Also adds a benchmark script for testing performance of the hidden states extractor. In a particular case of interest for GPT-OSS, I get a 1.45x improvement over the baseline, and am able to achieve within 10% of baseline throughput performance in some cases (with a lot of parameter tuning)

## Testing

Extensively validated in manual EAGLE training runs

### CODE REVIEW

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 1]
```diff
@@ -174,6 +260,26 @@ def register_kv_caches(self, kv_caches: dict[str, torch.Tensor]):
...
+    ) -> None:
+        """Thread worker: wait for async DtoH copy, write to disk, release lock.
+
+        ``lock_fd`` is an open file descriptor on the companion ``.lock``
+        file with ``LOCK_EX`` already held.  Closing it releases the lock,
+        which unblocks any client sleeping on ``LOCK_SH``.
+        """
+        try:
+            event.synchronize()
+            torch.save(tensors, filename)
```
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:278]: Why are you using torch.save instead of safetensors now? I think this would break existing integrations and uses pickle afaik

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:436]: This hasn't been updated, are we just getting rid of the cached_reqs path now?

File: examples/offline_inference/extract_hidden_states.py [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
+# update the benchmark in benchmarks/benchmark_hidden_state_extraction.py
+# and the docs in docs/features/speculative_decoding/extract_hidden_states.md
 
 # Example: Using the custom "extract_hidden_states" speculator method and
 # ExampleHiddenStatesConnector to extract and save hidden states from vllm
 
 with tempfile.TemporaryDirectory() as tmpdirname:
     llm = LLM(
         model="Qwen/Qwen3-8B",  # Your target model
+        enable_chunked_prefill=False,  # required
```
Comment (mgoin) [inline examples/offline_inference/extract_hidden_states.py:22]: Could we just make activating the `extract_hidden_states` method set this for the user?

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 3]
```diff
@@ -329,7 +470,31 @@ def request_finished(
...
+        return True, {"hidden_states_path": req_filename}
+
+    def get_finished(
+        self, finished_req_ids: set[str]
+    ) -> tuple[set[str] | None, set[str] | None]:
+        """Poll async write completion for requests that finished generating.
+
+        The scheduler passes finished_req_ids to tell the worker which
+        requests are done generating.  We accumulate these across calls
+        and return a request as "finished sending" once its disk write
```
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:482]: Does get_finished really need to wait for disk write? It seems to me that it could return after the d2h copy completes, after `copy_done.record(copy_stream)`

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 4]
```diff
@@ -39,6 +40,26 @@ def extract_from_kv_cache(
...
+        path: The file path returned in kv_transfer_params["hidden_states_path"].
+
+    Returns:
+        Dict with "hidden_states" and "token_ids" tensors.
+    """
+    lock_path = path + ".lock"
+    with open(lock_path) as lf:
+        fcntl.flock(lf, fcntl.LOCK_SH)  # sleeps until writer releases LOCK_EX
+        data = torch.load(path, map_location="cpu")
+    return data
```
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:60]: Should we add a cleanup util within the connector for these lock files? Currently they are implemented in both the benchmark and the example so I assume these are never cleaned up (not that big of a deal)

File: vllm/v1/spec_decode/extract_hidden_states.py [hunk 5]
```diff
@@ -299,18 +310,18 @@ def prepare_next_token_ids_padded(
...
-        backup_tokens_gpu = torch.tensor(
+        self.backup_next_token_ids.np[:num_reqs] = np.array(
             [
                 requests[gpu_input_batch.req_ids[i]].get_token_id(seq_lens_list[i])
                 for i in range(num_reqs)
             ],
-            dtype=torch.int32,
-            device=device,
+            dtype=np.int32,
         )
```
Comment (mgoin) [inline vllm/v1/spec_decode/extract_hidden_states.py:322]: nit: I don't think you need the `np.array` here as it can copy from a list directly

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 1 above)
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:278]: Performance is much better using torch.save for some reason. I expected safetensors to perform better, but in multiple benchmarks the safetensors implementation seems to hold the GIL and bottleneck parallel writing. Torch.save seems to release it properly.

Safetensors claims to release it, so it might be a bug. I did pretty extensive profiling to come to this conclusion.

File: examples/offline_inference/extract_hidden_states.py (hunk 2 above)
Comment (benchislett) [reply examples/offline_inference/extract_hidden_states.py:22]: I tried this first, but it seems tricky as a lot of the vllm initialization is downstream of the scheduler config init. I could have the scheduler config post_init check if the extract hidden states connector is used, but that seems pretty ugly.

Do you know if there's a better way to handle this? It's possible I'm just naive here

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:436]: oops, my bad. yeah I'll remove this

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 4 above)
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:60]: yeah, will do. I implemented this at some point but lost track of the diff

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 3 above)
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:482]: This is a good point. For some reason I thought we couldn't distinguish between the two, but since we run the copies from the main thread and only use the thread pool for the file writes, we can just wait for the DtoH instead of the lifetime of the writer thread.

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 1 above)
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: Has now been fixed in safetensors 0.8, will switch back and we'll pick it up soon
https://github.com/safetensors/safetensors/issues/541

Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: ## Performance with safetensors 0.8.0rc0

```
============================================================
  BASELINE RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           11.43s
  Tokens/s:            183,480
  Prompts/s:           44.80
============================================================

============================================================
  EXTRACT RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           11.54s
  Tokens/s:            181,715
  Prompts/s:           44.36
  Mean hidden value:   0.015989
============================================================

Extraction slowdown factor: 1.01x
```

## Performance with safetensors 0.7.0 (latest release)

```
============================================================
  BASELINE RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           11.43s
  Tokens/s:            183,402
  Prompts/s:           44.78
============================================================

============================================================
  EXTRACT RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           20.49s
  Tokens/s:            102,334
  Prompts/s:           24.98
  Mean hidden value:   0.015990
============================================================

Extraction slowdown factor: 1.79x
```

## Performance with torch.save

```
============================================================
  BASELINE RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           11.40s
  Tokens/s:            184,022
  Prompts/s:           44.93
============================================================

============================================================
  EXTRACT RESULTS
============================================================
  Prompts:             512
  Total prompt tokens: 2,097,152
  Wall time:           11.54s
  Tokens/s:            181,734
  Prompts/s:           44.37
  Mean hidden value:   0.015989
============================================================

Extraction slowdown factor: 1.01x
```

**Benchmark command:**
```
python benchmarks/benchmark_hidden_state_extraction.py \
    --model Qwen/Qwen3-0.6B \
    --num-prompts 512 \
    --num-clients 8 \
    --prompt-len 4096 \
    --layers 1 4 6 11
```

on RTX Pro 6000 Blackwell

Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: Updated to use safetensors and print a warning if the version is < 0.8.0

Comment (mgoin) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: Wow, nice find indeed

File: examples/features/speculative_decoding/extract_hidden_states_offline.py [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
+# update the benchmark in benchmarks/benchmark_hidden_state_extraction.py
+# and the docs in docs/features/speculative_decoding/extract_hidden_states.md
 
 # Example: Using the custom "extract_hidden_states" speculator method and
 # ExampleHiddenStatesConnector to extract and save hidden states from vllm
 
 with tempfile.TemporaryDirectory() as tmpdirname:
     llm = LLM(
         model="Qwen/Qwen3-8B",  # Your target model
+        enable_chunked_prefill=False,  # required
```
Comment (mgoin) [reply examples/features/speculative_decoding/extract_hidden_states_offline.py:21]: Hmm I would assume you could put it in vllm_config post-init for cross-config features but would need to check

Comment (mgoin) [reply examples/features/speculative_decoding/extract_hidden_states_offline.py:21]: Yeah I think here is a good place already https://github.com/vllm-project/vllm/blob/8eb401134e750781a202c0b6dc4059616cdb4954/vllm/config/vllm.py#L676-L681
something like
```python
if (
    self.kv_transfer_config is not None
    and self.kv_transfer_config.kv_connector == "ExampleHiddenStatesConnector"
    and self.scheduler_config.enable_chunked_prefill
):
    logger.info(
        "Disabling chunked prefill for ExampleHiddenStatesConnector."
    )
    self.scheduler_config.enable_chunked_prefill = False
```

File: vllm/config/vllm.py [hunk 7]
```diff
@@ -679,6 +679,17 @@ def _post_init_kv_transfer_config(self) -> None:
...
+        # Check if KV connector requires chunked prefill to be disabled.
+        if (
+            self.kv_transfer_config is not None
+            and self.kv_transfer_config.kv_connector == "ExampleHiddenStatesConnector"
+            and self.scheduler_config.enable_chunked_prefill
+        ):
+            raise ValueError(
+                "ExampleHiddenStatesConnector does not support chunked prefill. "
+                "Please disable chunked prefill (--no-enable-chunked-prefill)."
+            )
```
Comment (mgoin) [inline vllm/config/vllm.py:745]: Oh I didn't see, why not just override the config here instead of failing?

**Review by mgoin — APPROVED** [review]
LGTM!

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 8]
```diff
@@ -125,17 +171,97 @@ def __init__(
...
+        for tensors, event, filename, req_id in self._pending_copies:
+            prior = self._req_futures.get(req_id)
+            assert prior is None, "Found another KV transfer request with same req_id!"
+
+            lock_fd = None
+            if self.use_lock:
+                # Create/open the lock file and acquire an exclusive lock.
+                # The lock is held by this fd; the thread worker will close
+                # the fd after writing, which releases the lock.
+                lock_path = filename + ".lock"
```
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:244]: A somewhat reasonable concern, but low priority for now
